scripts/o3de/tests/test_enable_gem.py

```diff
@@ -164,3 +190,152 @@ def add_gem_dependency(enable_gem_cmake_file: pathlib.Path, gem_name: str):
+        pytest.param('1.2.3', ['testgem1','testgem2'], { 'testgem1':'1.2.3', 'testgem2':'2.0.0'}, 
+                    False, False, ['o3de-test<=1.0.0'], [], 'o3de-test', '1.0.0', {}, 0),
+        # passes when dependent gems with and without version specifiers are used
+        pytest.param('1.2.3', ['testgem1','testgem2~=2.0.0'], { 'testgem1':'1.2.3', 'testgem2':'2.0.1'}, 
+                    False, False, ['o3de-test<=1.0.0'], [], 'o3de-test', '1.0.0', {}, 0),
+        # fails when dependent gem is not found
+        pytest.param('1.2.3', ['testgem1==1.2.3','testgem2>1.0.0'], { 'testgem1':'1.2.3'}, 
+                    False, False, ['o3de-test<=1.0.0'], [], 'o3de-test', '0.0.0', {}, 1),
+        # fails when dependent gem with wrong version found
+        pytest.param('1.2.3', ['testgem1==1.2.3','testgem2>1.0.0'], { 'testgem1':'1.0.0', 'testgem2':'1.0.0'}, 
+                    False, False, ['o3de-test~=1.0.0'], [], 'o3de-test', '1.0.0', {}, 1),
+        # does not modify project when check only 
+        pytest.param('1.2.3', ['testgem1==1.2.3','testgem2>1.0.0'], { 'testgem1':'1.2.3', 'testgem2':'2.0.0'}, 
+                    True, False, ['o3de-test<=1.0.0'], [], 'o3de-test', '1.0.0', {}, 0),
+        # passes when a engine api versions found
+        pytest.param('1.2.3', [], {}, False, False, [], ['api==1.2.3'], 'o3de-test', '1.0.0', {'api':'1.2.3'}, 0),
+        # passes when a engine api without a version specifier is used and api exists 
+        pytest.param('1.2.3', [], {}, False, False, [], ['api'], 'o3de-test', '1.0.0', {'api':'1.2.3'}, 0),
+        # fails when engine api not found 
+        pytest.param('1.2.3', [], {}, False, False, [], ['api==2.3.4'], 'o3de-test', '1.0.0', {'otherapi':'1.2.3'}, 1),
+        # fails when engine api version incorrect
+        pytest.param('1.2.3', [], {}, False, False, [], ['api==2.3.4'], 'o3de-test', '1.0.0', {'api':'1.2.3'}, 1),
+        # passes when engine matches compatible_engines but engine api version doesn't match 
+        pytest.param('1.2.3', [], {}, False, False, ['o3de-test>=1.0.0'], ['api==1.2.3'], 'o3de-test', '2.0.0', {'api':'2.3.4'}, 0),
+        # passes when engine version doesn't match but api version matches
+        pytest.param('1.2.3', [], {}, False, False, ['o3de-test==1.0.0'], ['api~=1.0'], 'o3de-test', '2.0.0', {'api':'1.9.2'}, 0),
+        ],
+    )
+    def test_enable_gem_checks_engine_compatibility(self, gem_version, gem_dependencies, 
+                                                    gem_names_and_versions, check, force, compatible_engines, 
+                                                    engine_api_dependencies, test_engine_name, test_engine_version, 
+                                                    test_engine_api_versions, expected_result):
+        gem_registered_with_project = True
+        gem_registered_with_engine = False
+        project_path = pathlib.PurePath('TestProject')
```

Comment:
There is still the case where the gem is registered with the user's global manifest (`~/.o3de_manifest.json`) instead of a project or engine that should be checked as well.